docs/adr/adr-006-fraud-service.md

```diff
@@ -0,0 +1,168 @@
+   Shares []*Share
+   Position uint8
+   isRow bool
+}
+
+type Share struct {
+   Share []byte
+   Proof nmt.Proof
+}
+```
+
+In addition, `das.Daser`:
+
+1. Creates a BEFP:
+
+```go
+const (
+   BadEncoding FraudProofType = "BadEncoding"
+)
+
+type BadEncoding struct {
+   Type FraudProofType
+   Height uint64
+   Shares []*Share
+   Position uint8
+   isRow bool
+}
+```
+
+2. Notify all light nodes via separate sub-service via proto message:
+
+```proto3
+
+message MerkleProof {
+  int64          total     = 1;
+  int64          index     = 2;
+  bytes          leaf_hash = 3;
+  repeated bytes aunts     = 4;
+}
+
+message Share {
+   bytes Share = 1;
+   MerkleProof Proof = 2;
+}
+
+message BadEnconding {
+   required string Type = 1;
+   required uint64 Height = 2;
+   repeated Share Shares = 3;
+   uint8 Position = 4;
+   bool isRow = 5;
+}
+```
+Data serialization/deserialization will be performed with `protobuf.Marshal`/`protobuf.Unmarshal`.
+
+`das.Daser` imports a data structure that implements `proof.FraudNotifier` interface that uses libp2p.pubsub under the hood:
+
+```go
+// FraudNotifier is a generic interface that sends a different kinds of fraud proofs to all subscribed on particular topic nodes
+type FraudNotifier interface {
+   // Broadcast takes a Fraud proof data stucture that implements standart BinaryMarshal interface and sends data to light nodes using libp2p pub-sub under the hood.
+   Notify(ctx context.Context, p Proof)  
+}
+```
+
+```go
+// FraudProofType is a enum type that represents particular fraud proof
+type FraudProofType string
+
+// Proof is a generic interface that will be used for all types of fraud proofs
+type Proof interface {
+   Type() FraudProofType
+   Height() (uint64, error)
+   Validate(*header.ExtendedHeader) error
+
+   encoding.BinaryMarshaller
+   encoding.BinaryUnmarshaller
+}
+```
+
+From the other side, light nodes have the ability to subscribe to a particular fraud proof update and verify received data:
+
+```go
+// Subscriber encompasses the behavior necessary to
+// subscribe/unsubscribe from new FraudProofs events from the
+// network.
+type Subscriber interface {
+   // Subscribe allows to subscribe on pub sub topic by it's type
+   Subscribe(ctx context.Context, proofType FraudProofType) (Subscription, error)
+}
+```
+
+```go
+// Subscription provides only valid proofs, because the validation is done under the hood of pub-sub
+type Subscription interface {
+   Proof() (Proof, error)
+}
+
+type FraudSub struct {
+   pubsub *pubsub.PubSub 
+}
+
+func NewFraudSub(p *pubsub.PubSub)(Subscription, error){}
+func(s *FraudSub) Proof() (Proof, error){}
+```
+
+```go
+// NOTE: re-think how FraudService should be designed(and constructed) for full nodes and for light nodes
+type FraudService struct {
+   badEncodingSub Subscription
+   fraudNotifier FraudNotifier
+}
+
+func(f *FraudService) Subscribe(ctx context.Context, proofType FraudProofType) (Subscription, error){}
```

Comment:
fraud service doesn't implement `Subscriber`, it implements `Subscription`.

So it needs these methods:

``` 
Proof()
Cancel()
```

as they're described on #L132

Reply:
added